mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td

```diff
@@ -42,6 +42,33 @@ def SYCL_Dialect : Dialect {
     /// first.
     llvm::Optional<llvm::StringRef> findMethod(::mlir::TypeID Type,
                                                llvm::StringRef Name) const;
+
+    /// Returns the name of the operation implementing the queried
+    /// method, if present.
+    ///
+    /// For a method to be queried, it must have been registered
+    /// first.
+    ///
+    /// This function is similar to findMethod(), but also works for
+    /// methods registered as members of classes derived from \p BaseType.
+    llvm::Optional<llvm::StringRef>
+    findMethodFromBaseClass(::mlir::TypeID BaseType,
+                            llvm::StringRef Name) const;
+
+
+    /// Add a definition for the given method to be used when lowering
+    /// SYCLMethodOpInterface instances.
+    ///
+    /// A call to this function will fail if \p MayOverride is false and we try to
```

Comment:
C++ member functions should have unique names, why do we need the functionality to override the mangled name?

Reply:
`cgeist` will first insert the function declaration (in `getOrCreateMLIRFunction`) and then, its definition (in `init`). We require via an assert that, when overriding, the mangled function name does not change.

Reply:
Instead of having `MayOverride`, can we override by checking if the given `FuncOp` is a declaration or definition? 
i.e., Exist in map -> assert name unchanged -> if `FuncOp` in map is declaration and new `FuncOp` is definition, then override.